internal/expr/prepare.go

```diff
@@ -41,7 +82,91 @@ func prepareInput(ti typeNameToInfo, p *inputPart) error {
 	return nil
 }
 
-// Prepare takes a parsed expression and all the Go objects mentioned in it.
+// prepareOutput checks that the output expressions are correspond to a known types.
+// It then checks they are formatted correctly and finally generates the columns for the query.
+func prepareOutput(ti typeNameToInfo, p *outputPart) ([]fullName, error) {
+
+	var outCols []fullName = make([]fullName, 0)
+
+	// Check target struct type and its tags are valid.
+	var info *info
+	var ok bool
+
+	for i, t := range p.target {
+		if i == 0 {
+			info, ok = ti[t.prefix]
+			if !ok {
+				return nil, fmt.Errorf(`type %s unknown, have: %s`, t.prefix, strings.Join(getKeys(ti), ", "))
+			}
+		} else if t.prefix != info.structType.Name() {
+			return nil, fmt.Errorf("multiple target types in: %s", p.String())
+		}
+
+		_, ok = info.tagToField[t.name]
+		if !ok && t.name != "*" {
+			return nil, fmt.Errorf(`type %s has no %q db tag`, info.structType.Name(), t.name)
+		}
+	}
+
+	if err := starCheckOutput(p); err != nil {
+		return nil, err
+	}
+
+	// Generate columns to inject into SQL query.
+
+	// Case 1: Star target cases e.g. "...&P.*".
+	if p.target[0].name == "*" {
+		info, _ := ti[p.target[0].prefix]
+
+		// Case 1.1: Single star e.g. "t.* AS &P.*" or "&P.*"
+		if len(p.source) == 0 || p.source[0].name == "*" {
+			pref := ""
+
+			// Prepend table name. E.g. "t" in "t.* AS &P.*".
+			if len(p.source) > 0 {
+				pref = p.source[0].prefix
+			}
+
+			tags := getKeys(info.tagToField)
+			for _, tag := range tags {
+				outCols = append(outCols, fullName{prefix: pref, name: tag})
+			}
+			return outCols, nil
+		}
+
+		// Case 1.2: Explicit columns e.g. "(col1, t.col2) AS &P.*".
+		if len(p.source) > 0 {
+			for _, c := range p.source {
+				if _, ok := info.tagToField[c.name]; !ok {
+					return nil, fmt.Errorf(`type %s has no %q db tag`, info.structType.Name(), c.name)
+				}
+				outCols = append(outCols, c)
+			}
+			return outCols, nil
+		}
+	}
+
+	// Case 2: None star target cases e.g. "...&(P.name, P.id)".
```

Comment:
Typo in the comment I think (`(&P...` vs `&(P...`). Same below.

Reply:
Fixed.